src/io-queue.c

```diff
@@ -0,0 +1,245 @@
+  }
+
+  // Mark responses as read in the ring.
+  const struct io_uring_cq *cq = &ring->cq;
+  io_uring_smp_store_release(cq->khead, *cq->khead + count);
+
+  assert(queue->inflightCount >= count);
+  queue->inflightCount -= count;
+
+  return count;
+}
+
+// Cancel all given requests.
+static inline void cancel_requests (XcpIoReq *reqs, int err) {
+  while (reqs) {
+    XcpIoReq *nextReq = STAILQ_NEXT(reqs, pImpl.next);
+    reqs->cb(reqs, err, reqs->userData);
+    reqs = nextReq;
+  }
+}
+
+// Fill a io_uring_sqe instance from a XcpIoReq.
+static inline void set_sqe_from_req (const XcpIoReq *req, struct io_uring_sqe *sqe) {
+  sqe->opcode = req->opcode == XcpIoOpcodeRead || req->opcode == XcpIoOpcodeReadV
+    ? IORING_OP_READV
+    : IORING_OP_WRITEV;
+  sqe->flags = 0;
+  sqe->ioprio = 0;
+  sqe->fd = req->fd;
+  sqe->off = (uint64_t)req->offset;
+  sqe->addr = (uint64_t)&req->iov;
+  sqe->len = req->opcode == XcpIoOpcodeRead || req->opcode == XcpIoOpcodeWrite ? 1 : (uint32_t)req->iov.iov_len;
+  sqe->rw_flags = 0;
+  sqe->user_data = (uint64_t)req;
+  sqe->__pad2[0] = sqe->__pad2[1] = sqe->__pad2[2] = 0;
+  sqe->buf_index = 0;
+}
+
+// -----------------------------------------------------------------------------
+
+int xcp_io_queue_init (XcpIoQueue *queue, size_t capacity, bool usePolling) {
+  // 1. Init fields.
+  memset(queue, 0, sizeof *queue);
+  if (!capacity)
+    return -EINVAL;
+
+  queue->eventFd = -1;
+  queue->usePolling = usePolling;
+
+  STAILQ_INIT(&queue->reqs);
+
+  int err = 0;
+
+  // 2. Create an eventfd to be notified when a request ends.
+  if (!usePolling && (queue->eventFd = eventfd(0, 0)) < 0)
+    return -errno;
+
+  // 3. Init ring.
+  struct io_uring *ring = &queue->pImpl.ring;
+  if ((err = io_uring_queue_init((unsigned int)capacity, ring, usePolling ? IORING_SETUP_IOPOLL : 0)) < 0) {
+    close(queue->eventFd);
+    queue->eventFd = -1;
+  } else if (!usePolling && (err = io_uring_register_eventfd(ring, queue->eventFd)) < 0)
+    xcp_io_queue_uninit(queue);
+  else
+    queue->capacity = capacity;
+
+  return err;
+}
+
+void xcp_io_queue_uninit (XcpIoQueue *queue) {
+  if (!queue->capacity)
+    return;
+
+  if (queue->eventFd != -1) {
+    close(queue->eventFd);
+    queue->eventFd = -1;
+  }
+  io_uring_queue_exit(&queue->pImpl.ring);
+}
+
+void xcp_io_queue_insert (XcpIoQueue *queue, XcpIoReq *req) {
+  STAILQ_INSERT_TAIL(&queue->reqs, req, pImpl.next);
+  ++queue->pendingCount;
+}
+
+int xcp_io_queue_submit (XcpIoQueue *queue) {
+  struct io_uring *ring = &queue->pImpl.ring;
+
+  // 1. Insert requests in the ring.
+  size_t n = 0;
+  XcpIoReq *req = STAILQ_FIRST(&queue->reqs);
+  if (XCP_LIKELY(req)) {
+    assert(queue->pendingCount);
+
+    struct io_uring_sqe *sqe = io_uring_get_sqe(ring);
+    if (XCP_LIKELY(sqe)) {
+      set_sqe_from_req(req, sqe);
+
+      for (++n; n < queue->pendingCount; ++n) {
+        struct io_uring_sqe *sqe = io_uring_get_sqe(ring);
+        if (!sqe)
```

Comment:
`XCP_UNLIKELY(sqe)`?

Reply:
No. I can't say if there is more than one free sqe in the ring.